File: 0045-jump-game-ii/0045-jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        """
        0 1 2 3 4
        1,2,3
            ^
            curr_end = 3            
            max_reach = 3  
            
        jumps = 1
        """
        if len(nums) == 1:
            return 0
        curr_end = nums[0]
        max_reach = nums[0]
        jumps = 1
        for i in range(len(nums) - 1):
            if i + nums[i] > max_reach:
                max_reach = i + nums[i]
            if i == curr_end:
                curr_end = max_reach
                jumps += 1

        return jumps


        
        

File: 0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        if (nums.length == 1) return 0;
        int jumps = 0;
        int currReach = nums[0], maxReach = nums[0];
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] + i > maxReach) maxReach = nums[i] + i;
            if (i == currReach || i == nums.length - 1) {
                currReach = maxReach;
                jumps++;
            }
        }
        return jumps;
    }
}